ObserverSet.swift

```diff
@@ -26,122 +26,198 @@
-  
-  // MARK: - Initialisers
-  
-  /**
-   Creates a new instance of an observer set.
-   
-   - returns: A new instance of an observer set.
-   */
-  public init() {}
-  
-  // MARK: - Public functions
-  
-  /**
-   Adds a new observer to the set.
-   
-   - parameter observer: The object that is to be notified.
-   - parameter callBack: The function to call on the observer when the notification is to be delivered.
-   
-   - returns: An entry in the list of observers, which can be used later to remove the observer.
-   */
-  @discardableResult
-  open func add<ObserverType: AnyObject>(_ observer: ObserverType, _ callBack: @escaping (ObserverType) -> (Parameters) -> Void) -> ObserverSetEntry<Parameters> {
-    let entry = ObserverSetEntry<Parameters>(observer: observer) { observer in callBack(observer as! ObserverType) }
-    synchronized {
-      self.entries.append(entry)
-    }
-    return entry
-  }
-  
-  /**
-   Adds a new function to the list of functions to invoke when a notification is to be delivered.
-   
-   - parameter callBack: The function to call when the notification is to be delivered.
-   
-   - returns: An entry in the list of observers, which can be used later to remove the observer.
-   */
-  @discardableResult
-  open func add(_ callBack: @escaping (Parameters) -> Void) -> ObserverSetEntry<Parameters> {
-    return self.add(self) { ignored in callBack }
-  }
-  
-  /**
-   Removes an observer from the list, using the entry which was returned when adding.
-   
-   - parameter entry: An entry returned when adding a new observer.
-   */
-  open func remove(_ entry: ObserverSetEntry<Parameters>) {
-    synchronized {
-      self.entries = self.entries.filter{ $0 !== entry }
-    }
-  }
-  
+	fileprivate var isExpired: Bool { false }
 
-  /**
-   Removes an observer from the list.
+	fileprivate func call(_ parameters: Parameters) {
+		// overridden
+	}
+}
 
-   - parameter observer: An observer to remove from the list of observers.
-   */
-  open func removeObserver(_ observer: AnyObject) {
-    synchronized {
-      self.entries = self.entries.filter{ $0.observer !== observer }
-    }
-  }
+private class ObserverSetEntryImpl<ObserverType: AnyObject, Parameters>: ObserverSetEntry<Parameters> {
+	
+	typealias ObserverCallback = (ObserverType) -> (Parameters) -> Void
+	
+	private(set) weak var observer: ObserverType?
+	private let callBack: ObserverCallback
+	private let queue: DispatchQueue?
+	
+	override var isExpired: Bool { observer == nil }
+	
+	init(queue: DispatchQueue?, observer: ObserverType, callBack: @escaping ObserverCallback) {
+		self.queue = queue
+		self.observer = observer
+		self.callBack = callBack
+	}
+	
+	override func call(_ parameters: Parameters) {
+		if let observer = observer {
+			let callBack = self.callBack(observer)
+			if let queue = queue {
+				queue.async {
+					callBack(parameters)
+				}
+			} else {
+				callBack(parameters)
+			}
+		}
```

Comment:
Very minor. But might be cleaner to guard around observer rather than have these nester `if let` statements. 

```suggestion
		guard let observer = observer else { return } 
		let callBack = self.callBack(observer)
		if let queue = queue {
			queue.async {
				callBack(parameters)
			}
		} else {
			callBack(parameters)
		}
```